File: scripts/github-actions/detect-delisted-stocks.ts
#!/usr/bin/env npx tsx
/**
 * 上場廃止銘柄を検出・削除するスクリプト
 *
 * 使用方法:
 *   npx tsx scripts/github-actions/detect-delisted-stocks.ts --dry-run  # 検出のみ
 *   npx tsx scripts/github-actions/detect-delisted-stocks.ts            # 検出して削除
 */

import { PrismaClient } from "@prisma/client"
import YahooFinance from "yahoo-finance2"

const prisma = new PrismaClient()
const yahooFinance = new YahooFinance()

interface DelistedStock {
  id: string
  tickerCode: string
  name: string
  reason: string
  usage: { portfolio: number; watchlist: number; transactions: number }
  canDelete: boolean
}

async function checkIfDelisted(tickerCode: string): Promise<{ isDelisted: boolean; reason: string }> {
  try {
    const quote = await yahooFinance.quote(tickerCode)

    // quoteがない場合は上場廃止
    if (!quote) {
      return { isDelisted: true, reason: "No quote data" }
    }

    // QuoteType=NONEは上場廃止
    if ((quote as { quoteType?: string }).quoteType === "NONE") {
      return { isDelisted: true, reason: "QuoteType=NONE" }
    }

    // symbolがない場合も上場廃止
    if (!quote.symbol) {
      return { isDelisted: true, reason: "No symbol in quote" }
    }

    // 現在価格がない場合は上場廃止の可能性
    if (!quote.regularMarketPrice) {
      return { isDelisted: true, reason: "No market price" }
    }

    return { isDelisted: false, reason: "Active" }
  } catch (error) {
    const errorMsg = String(error)
    if (errorMsg.toLowerCase().includes("possibly delisted") || errorMsg.toLowerCase().includes("not found")) {
      return { isDelisted: true, reason: `Delisted: ${errorMsg}` }
    }
    return { isDelisted: false, reason: `Error: ${errorMsg}` }
  }
}

async function getStockUsage(stockId: string): Promise<{ portfolio: number; watchlist: number; transactions: number }> {
  const [portfolio, watchlist, transactions] = await Promise.all([
    prisma.portfolioStock.count({ where: { stockId } }),
    prisma.watchlistStock.count({ where: { stockId } }),
    prisma.transaction.count({ where: { stockId } }),
  ])

  return { portfolio, watchlist, transactions }
}

function parseArgs(): { dryRun: boolean; limit: number } {
  const args = process.argv.slice(2)
  let dryRun = false
  let limit = 100

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--dry-run") {
      dryRun = true
    } else if (args[i] === "--limit" && args[i + 1]) {
      limit = parseInt(args[++i], 10)
    }
  }

  return { dryRun, limit }
}

async function main(): Promise<void> {
  const { dryRun, limit } = parseArgs()

  try {
    // 全銘柄を取得
    console.log(`Fetching stocks to check (limit: ${limit})...`)
    const stocks = await prisma.stock.findMany({
      orderBy: { createdAt: "asc" },
      take: limit,
      select: { id: true, tickerCode: true, name: true },
    })

    console.log(`Found ${stocks.length} stocks to check\n`)

    const delistedStocks: DelistedStock[] = []
    const activeStocks: { tickerCode: string; name: string }[] = []

    for (let idx = 0; idx < stocks.length; idx++) {
      const stock = stocks[idx]
      process.stdout.write(`[${idx + 1}/${stocks.length}] Checking ${stock.tickerCode} (${stock.name})... `)

      const { isDelisted, reason } = await checkIfDelisted(stock.tickerCode)

      if (isDelisted) {
        console.log(`❌ DELISTED (${reason})`)

        const usage = await getStockUsage(stock.id)
        const totalUsage = usage.portfolio + usage.watchlist + usage.transactions

        delistedStocks.push({
          id: stock.id,
          tickerCode: stock.tickerCode,
          name: stock.name,
          reason,
          usage,
          canDelete: totalUsage === 0,
        })
      } else {
        console.log(`✅ Active (${reason})`)
        activeStocks.push({
          tickerCode: stock.tickerCode,
          name: stock.name,
        })
      }
    }

    // レポート出力
    console.log("\n" + "=".repeat(60))
    console.log("SUMMARY")
    console.log("=".repeat(60))
    console.log(`Total checked: ${stocks.length}`)
    console.log(`Active stocks: ${activeStocks.length}`)
    console.log(`Delisted stocks: ${delistedStocks.length}`)

    if (delistedStocks.length > 0) {
      console.log("\n" + "=".repeat(60))
      console.log("DELISTED STOCKS")
      console.log("=".repeat(60))

      const canDelete = delistedStocks.filter((s) => s.canDelete)
      const cannotDelete = delistedStocks.filter((s) => !s.canDelete)

      if (canDelete.length > 0) {
        console.log(`\nCan delete (${canDelete.length} stocks):`)
        for (const stock of canDelete) {
          console.log(`  - ${stock.tickerCode}: ${stock.name}`)
          console.log(`    Reason: ${stock.reason}`)
        }
      }

      if (cannotDelete.length > 0) {
        console.log(`\nCannot delete - in use (${cannotDelete.length} stocks):`)
        for (const stock of cannotDelete) {
          console.log(`  - ${stock.tickerCode}: ${stock.name}`)
          console.log(
            `    Usage: Portfolio=${stock.usage.portfolio}, ` +
              `Watchlist=${stock.usage.watchlist}, ` +
              `Transactions=${stock.usage.transactions}`
          )
        }
      }

      // 削除実行
      if (!dryRun && canDelete.length > 0) {
        console.log("\n" + "=".repeat(60))
        console.log("DELETING STOCKS")
        console.log("=".repeat(60))

        for (const stock of canDelete) {
          process.stdout.write(`Deleting ${stock.tickerCode}... `)
          await prisma.stock.delete({ where: { id: stock.id } })
          console.log("✅ Done")
        }

        console.log(`\n✅ Successfully deleted ${canDelete.length} delisted stocks`)
      } else if (dryRun) {
        console.log("\n⚠️  DRY RUN mode - no stocks were deleted")
      }
    }
  } catch (error) {
    console.error(`\n❌ Error: ${error}`)
    process.exit(1)
  } finally {
    await prisma.$disconnect()
  }
}

main()

export {}

File: scripts/github-actions/detect_delisted_stocks.py
#!/usr/bin/env python3
"""上場廃止銘柄を検出・削除するスクリプト"""

import argparse
import os
import sys
import yfinance as yf
import psycopg2


def get_database_url() -> str:
    url = os.environ.get("DATABASE_URL")
    if not url:
        print("Error: DATABASE_URL environment variable not set")
        sys.exit(1)
    return url


def check_if_delisted(ticker_code: str) -> tuple[bool, str]:
    try:
        symbol = ticker_code if ticker_code.endswith(".T") else f"{ticker_code}.T"
        ticker = yf.Ticker(symbol)
        hist = ticker.history(period="1mo")
        if hist.empty:
            return True, "No price data available"
        return False, "Active"
    except Exception as e:
        error_msg = str(e).lower()
        if "delisted" in error_msg or "not found" in error_msg:
            return True, f"Delisted: {e}"
        return False, f"Error: {e}"


def get_stock_usage(conn, stock_id: str) -> dict:
    with conn.cursor() as cur:
        cur.execute('SELECT COUNT(*) FROM "PortfolioStock" WHERE "stockId" = %s', (stock_id,))
        portfolio = cur.fetchone()[0]
        cur.execute('SELECT COUNT(*) FROM "WatchlistStock" WHERE "stockId" = %s', (stock_id,))
        watchlist = cur.fetchone()[0]
        cur.execute('SELECT COUNT(*) FROM "Transaction" WHERE "stockId" = %s', (stock_id,))
        transactions = cur.fetchone()[0]
    return {"portfolio": portfolio, "watchlist": watchlist, "transactions": transactions}


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--limit", type=int, default=100)
    args = parser.parse_args()

    conn = psycopg2.connect(get_database_url())

    try:
        print(f"Fetching stocks to check (limit: {args.limit})...")
        with conn.cursor() as cur:
            cur.execute('SELECT id, "tickerCode", name FROM "Stock" ORDER BY "createdAt" ASC LIMIT %s', (args.limit,))
            stocks = cur.fetchall()

        print(f"Found {len(stocks)} stocks to check\n")

        delisted_stocks, can_delete = [], []

        for idx, (stock_id, ticker_code, name) in enumerate(stocks):
            print(f"[{idx + 1}/{len(stocks)}] Checking {ticker_code} ({name})... ", end="", flush=True)
            is_delisted, reason = check_if_delisted(ticker_code)

            if is_delisted:
                print(f"❌ DELISTED ({reason})")
                usage = get_stock_usage(conn, stock_id)
                total_usage = usage["portfolio"] + usage["watchlist"] + usage["transactions"]
                delisted_stocks.append({"id": stock_id, "tickerCode": ticker_code, "name": name, "reason": reason, "usage": usage, "canDelete": total_usage == 0})
                if total_usage == 0:
                    can_delete.append({"id": stock_id, "tickerCode": ticker_code})
            else:
                print(f"✅ Active")

        print(f"\n{'=' * 60}\nSUMMARY\n{'=' * 60}")
        print(f"Total checked: {len(stocks)}\nDelisted stocks: {len(delisted_stocks)}")

        if can_delete and not args.dry_run:
            print(f"\nDELETING {len(can_delete)} stocks...")
            with conn.cursor() as cur:
                for stock in can_delete:
                    cur.execute('DELETE FROM "Stock" WHERE id = %s', (stock["id"],))
                    print(f"  Deleted {stock['tickerCode']}")
            conn.commit()
            print(f"✅ Successfully deleted {len(can_delete)} delisted stocks")
        elif args.dry_run:
            print("\n⚠️  DRY RUN mode - no stocks were deleted")

    finally:
        conn.close()


if __name__ == "__main__":
    main()
